Add GssapiRequiredNameAttributes option

mrogers950 opened the pull request:
Allow setting the required name attributes for a session, using a flex and
bison based parser to process the attributes specified.

frozencemetery commented on Makefile.am:
```diff
@@ -1,14 +1,17 @@
+check: all check-na
 	cd $(srcdir) && ./tests/magtests.py --path $(TESTSDIR) --so-dir=$(abs_builddir)/src/.libs
 
+check-na:
```
Maybe just fold this into the main `check` target, rather than putting it on its own?

mrogers950 replied:
I can, that was just for me to be able to run the shell tests separately.

frozencemetery commented on configure.ac:
```diff
@@ -83,14 +86,14 @@ AC_CHECK_FUNCS(gss_krb5_ccache_name)
 AC_SUBST([GSSAPI_LIBS])
 
-MAG_CFLAGS="`${APXS} -q CFLAGS` `${APXS} -q EXTRA_CPPFLAGS` `${APR} --cflags` ${GSSAPI_CFLAGS} ${OPENSSL_CFLAGS} -I`${APXS} -q INCLUDEDIR` `${APR} --includes`"
+MAG_CFLAGS="`${APXS} -q CFLAGS` `${APXS} -q EXTRA_CPPFLAGS` `${APR} --cflags` ${GSSAPI_CFLAGS} ${OPENSSL_CFLAGS} -I`${APXS} -q INCLUDEDIR` `${APR} --includes` -fPIC"
```
Adding PIE/PIC support is a fine thing to do, but not in scope for this PR.

mrogers950 replied:
I forgot to comment that building fails without adding -fPIC here;
/usr/bin/ld: lex/libparser.a(libparser_a-lex.o): relocation R_X86_64_PC32 against undefined symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value

I can add it to a separate commit before the main one (and submit a separate PR if you want)

frozencemetery replied:
Ah, that makes sense.  Maybe just drop a comment above it.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+                        continue;
+                    }
+                    if (($3[0] == '[') && ($3[strlen($3) - 1] == ']')) {
+                        if (hex2bincmp($3 + 1, strlen($3) - 2, (unsigned char *)vals[i], strlen(vals[i]))) {
```
Wrap at <80 characters, please.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+
+/* Convert hex_len of hex characters into binary and memcmp against bin. Return
+ * 1 if the hex string is valid and matches, else 0. */
+int hex2bincmp(const char *hex, size_t hex_len, unsigned char *bin, size_t bin_len)
```
<80 here too.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+    r = memcmp(b, bin, b_len);
+    free(b);
+
+    return r == 0 ? 1 : 0;
```
Can be simplified to just `return r == 0`.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+        return 1;
+    }
+
+    /* No name attributes but required attributes
```
This should be re-wrapped.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+    ret = yyparse(NULL, NULL, NULL);
+    yy_delete_buffer(buffer);
+
+    return (ret == 0) ? 1 : 0;
```
Just `return ret == 0` as above.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,205 @@
+
+void yyerror(const char **keys, const char **vals, int *status, const char *s)
+{
+    return;
```
Log something here?

mrogers950 replied:
My understanding is that you always have to define a yyerror() function and it's called on a syntax error, which we handle with an apache log message during expect_name_attrs().

frozencemetery replied:
Could do a comment for that, or not.  Thanks for explaining.

simo5 replied:
please prepend functions that are not static with a mag_ prefix

frozencemetery commented on src/mod_auth_gssapi.c:
```diff
@@ -1151,6 +1151,30 @@ static int mag_auth(request_rec *req)
+     * expected_name_attr_verify(). */
+    names = apr_pcalloc(req->pool, (mc->na_count + 1) * sizeof(*names));
+    vals = apr_pcalloc(req->pool, (mc->na_count + 1) * sizeof(*names));
+    for (i = 0; i < mc->na_count; i++) {
```
Please use C99 `for`-loop syntax.

frozencemetery commented on src/mod_auth_gssapi.h:
```diff
@@ -138,3 +139,6 @@ struct mag_conn {
 const char *mag_str_auth_type(int auth_type);
 char *mag_error(apr_pool_t *pool, const char *msg, uint32_t maj, uint32_t min);
+extern int expected_name_attr_verify(const char *expr, const char **attrs,
+			      const char **vals);
```
`const char **vals` should line up with `const char *expr`.

frozencemetery commented on tests/natest.c:
```diff
@@ -0,0 +1,62 @@
+        exit(1);
+    }
+
+    for (i = 2; i < argc; i++) {
```
C99 `for`-loops please.

frozencemetery commented on tests/natest.sh:
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash -xe
+cmd=./natest
+
+  # Single match and wildcard.
```
Could these be echoed instead of just being comments?  Might make the logs easier to read through.

frozencemetery commented on src/environ.c:
```diff
@@ -291,6 +291,12 @@ void mag_get_name_attributes(request_rec *req, struct mag_config *cfg,
+            mc_add_name_attribute(mc, apr_pstrndup(req->pool,
+                                                   attr.name.value,
+                                                   attr.name.length),
+                                  apr_pstrndup(req->pool,
```
This is kind of gross looking.  Can you pull out the `apr_pstrndup` calls into variables?

simo5 commented on src/mod_auth_gssapi.c:
```diff
@@ -1795,6 +1845,8 @@ static const command_rec mag_commands[] = {
                     "Don't resend negotiate header on negotiate failure"),
     AP_INIT_RAW_ARGS("GssapiNameAttributes", mag_name_attrs, NULL, OR_AUTHCFG,
                      "Name Attributes to be exported as environ variables"),
+    AP_INIT_RAW_ARGS("GssapiExpectedNameAttributes", expect_name_attrs, NULL,
```
nitpick: if they are Required ... then the name should be GssapiRequiredNameAttributes ?

mrogers950 replied:
I think that might actually be the better name now that you mention it.

frenche commented on README:
```diff
@@ -306,6 +306,30 @@ underscores for environment variable names.
+#### Example
+    GssapiExpectedNameAttributes "auth-indicators=high"
+    GssapiExpectedNameAttributes "auth-indicators=high or other-attr=foo"
+    GssapiExpectedNameAttributes "((auth-indicators=low and auth-indicators=med) or auth-indicators=high)"
```
Can auth-indicators be both?

mrogers950 replied:
Yes, you can have multiple authentication indicators (although that may not be clear with the given example), and the parser will use the list of them when checking for the match on each attribute in the statement.

frenche replied:
Hmm, how are we handling multiple values? from the code it looks like we add _N to the name, is that right? does it parse ok?

frenche replied:
Oh, I now see that you added a copy of each value.

frenche commented on src/mod_auth_gssapi.c:
```diff
@@ -1693,6 +1728,21 @@ static const char *mag_name_attrs(cmd_parms *parms, void *mconfig,
+    if (!expected_name_attr_expr_check(w)) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, parms->server,
+                     "Invalid Name Attribute filter [%s].", w);
+        return NULL;
```
I think we should return no null here, so it won get silently ignored.

mrogers950 commented:
I've turned "Expected" to "Required", revised the README text, and now return an error string from expect_name_attrs() (now required_name_attrs()) on a syntax error.

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,209 @@
+extern int yyparse(const char **keys, const char **vals, int *status);
+extern YY_BUFFER_STATE yy_scan_string(char * str);
+extern void yy_delete_buffer(YY_BUFFER_STATE buffer);
+int hex2bincmp(const char *hex, size_t hex_len, unsigned char *bin,
```
`bin` and `bin_len` on their own line; line up with `hex`

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,209 @@
+requiredkv: STRING EQUAL STRING {
+            int i, ret = 0;
+            if (keys != NULL && vals != NULL) {
+                for (i = 0; keys[i] != NULL && vals[i] != NULL; i++) {
```
C99

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,209 @@
+          | STRING EQUAL AST {
+            int i, ret = 0;
+            if (keys != NULL && vals != NULL) {
+                for (i = 0; keys[i] != NULL && vals[i] != NULL; i++) {
```
C99

frozencemetery commented on src/lex/parser.y:
```diff
@@ -0,0 +1,209 @@
+
+static int hexchar(unsigned int c)
+{
+    if ((c >= '0') && (c <= '9'))
```
Parens not needed; also applies to rest of function

frenche commented:
Do attributes specified in GssapiRequiredNameAttributes need to be specified in GssapiNameAttributes too? from the code it looks like yes, so maybe we need to clarify it, or address it in code.

simo5 commented:
Uhmm good point @frenche I wonder if we should error out if the required is not a subset or if we should automaticaly add the ones in the required parser (do we have a simple way to get a list of them) to the set of items we retrieve, or if we should "shadow" retrieve them.
We may not want to expose named attributes we use for authorization to the application necessarily.

simo5 commented on src/environ.c:
```diff
@@ -292,6 +294,11 @@ void mag_get_name_attributes(request_rec *req, struct mag_config *cfg,
+            v = apr_pstrndup(req->pool, attr.value.value, attr.value.length);
+
+            mc_add_name_attribute(mc, n, v);
+
```
so I have been thinking about this.
1) Why are you forcibly re-adding the name attribute here ? This will break the semantics of the Name Attribute interface I think.
2) do not use 1 letter vars
3) in a case like this I think it makes more sense to declare these var in the do{} block, not across the whole function
4) I think we need separate lists for required_name_attributes

This last point is the main one, I think we should have a separate list for required attributes because some people may want to use specific name attribute (say auth indicators) without being forced to expose them as env vars to the application.
So I guess we need a required_nameattr variable on mc where we set attributes use to check required attributes.

mrogers950 replied:
Sure, I see how this is problematic. I'll separate the required attributes from the env exposed ones. As for the smoke test, against a 1.16 krb5 KDC we could enable the encrypted challenge indicator (does not require as much config as PKINIT/OTP) and test against that, otherwise I'm not sure what other name attribute we could enable in the test suite and test against.

simo5 replied:
Any name attribute will work, pick the simplest to expose

frenche replied:
+1 on the above.
One way to go about it could be to store raw attributes on the MC struct (if we need it for either env or authz), and defer the parsing to a later point.

simo5 commented on src/mod_auth_gssapi.c:
```diff
@@ -1151,6 +1151,30 @@ static int mag_auth(request_rec *req)
 
+/* Return 1 if the received name attributes are allowed by the
+ * GssapiRequiredNameAttributes option, or else 0. */
+static int verify_name_attributes(struct mag_config *cfg, request_rec *req,
```
-> verify_required_name_attributes()

simo5 commented on src/mod_auth_gssapi.c:
```diff
@@ -1151,6 +1151,30 @@ static int mag_auth(request_rec *req)
+    /* Use NULL terminated string arrays to pass name attributes and values to
+     * required_name_attr_verify(). */
+    names = apr_pcalloc(req->pool, (mc->na_count + 1) * sizeof(*names));
+    vals = apr_pcalloc(req->pool, (mc->na_count + 1) * sizeof(*names));
```
As you change this code to store attributes into a separate required_nameattr structure, store them there already zero terminated. Please also check that the values do not contain embedded zeros by verifying with strlen() that the length match what you expect and fire a debug log if they don't. (Or perhaps auto-base64 the value in this case ?)

simo5 replied:
I guess if you do the above this function can go away entirely ...

simo5 commented on src/mod_auth_gssapi.c:
```diff
@@ -1795,6 +1845,8 @@ static const command_rec mag_commands[] = {
     AP_INIT_RAW_ARGS("GssapiNameAttributes", mag_name_attrs, NULL, OR_AUTHCFG,
                      "Name Attributes to be exported as environ variables"),
+    AP_INIT_RAW_ARGS("GssapiRequiredNameAttributes", required_name_attrs, NULL,
+                     OR_AUTHCFG, "Name Attributes required to be present"),
```
If we support != expressions then this description is not correct.
We should say something like: Required checks for Name Attributes

simo5 replied:
and perhaps the option should actually be called GssapiCheckNameAttributes ? Or GssapiVerifyNameAttributes now that I think of it ? Sorry for not realizing this sooner.

simo5 replied:
Seems we do not support !=, is this a conscious decision to avoid "deny" accerss control cases ?
If not what is the reasoning ?
I am perfectly fine *not* supporting !=, but I think we should add to the README explicitly that we match only on '=' and we do not support '!='

mrogers950 replied:
No particular reason, though we certainly can support deny rules if you think there will be demand for it.

simo5 replied:
I guess someone will ask, but in general I tend to be against the concept of "deny ACL" because they do not fail safe.

simo5 commented on src/mod_auth_gssapi.h:
```diff
@@ -138,3 +139,6 @@ struct mag_conn {
 char *mag_error(apr_pool_t *pool, const char *msg, uint32_t maj, uint32_t min);
+extern int required_name_attr_verify(const char *expr, const char **attrs,
+                                     const char **vals);
+extern int required_name_attr_expr_check(const char *expr);
```
Is there a reason why these are not being pulled in with a dedicated include file ?

mrogers950 replied:
I'll make a dedicated include for these.

mrogers950 commented:
I've pushed my current changes, that should be mostly everything except for the smoke test which I am working on, and will push later.

simo5 commented on src/environ.c:
```diff
@@ -218,6 +264,51 @@ static void mag_add_json_name_attr(request_rec *req, bool first,
+            if (!mag_get_name_attr(req, name, &attr)) {
+                break;
+            }
+            mc_add_req_name_attribute(req, mc, attr.name, attr.value);
```
In the oriignal name attribute code we pull in display_value and fall back to value only if that is not available.

simo5 commented on src/environ.c:
```diff
@@ -43,6 +43,52 @@ static apr_status_t mag_mc_name_attrs_cleanup(void *data)
+                                                  name.length);
+    mc->required_name_attrs[count + 1] = NULL;
+
+    mc->required_name_vals[count] = apr_pstrndup(mc->pool, value.value,
```
There is no check that value is in fact a string, this code will not handle binary values with embedded zeros well if such are passed in.
Or are we sure name attributes are always guaranteed to be printable strings ?

simo5 commented on src/environ.c:
```diff
@@ -368,6 +459,19 @@ void mag_export_req_env(request_rec *req, apr_table_t *env)
+
+    ap_set_module_config(req->request_config, &auth_gssapi_module, mc->env);
+    mag_export_req_env(req, mc->env);
+}
```
Should we just set the error env var here and then call mag_set_req_data() to complete ?
It seem like maiing env vars and ccaches available to the error handler may be valuable given yiou are calling mag_export_req_env() here but not mag_set_name_attributes() which is somewhat required to fill the proper tables.

simo5 commented on src/environ.c:
```diff
@@ -368,6 +459,19 @@ void mag_export_req_env(request_rec *req, apr_table_t *env)
+                           struct mag_conn *mc)
+{
+    apr_table_set(mc->env, "GSS_NAME", mc->gss_name);
+    apr_table_set(mc->env, "GSS_NAME_ATTR_ERROR",
```
Should this be GSS_REQUIRED_NAME_ATTR_ERROR ?

simo5 commented on src/mod_auth_gssapi.c:
```diff
@@ -1176,6 +1177,20 @@ static int mag_complete(struct mag_req_cfg *req_cfg, struct mag_conn *mc,
     mc->expiration = time(NULL) + vtime;
 
     mag_get_name_attributes(req, cfg, client, mc);
+    mag_get_required_name_attributes(req, cfg, client, mc);
```
I understand why you spoon this out in a new function, however, this means we call gss_get_name_attribute twice for each attribute, and it may be an expensive call.
Can we rather restructure  mag_get_name_attributes() so it calls gss_get_name_attribute() *once but then internally calls two subfunctions to fill the two arrays ?

EDIT: added *once

mrogers950 commented:
Sorry about the delay on this update. Here's an overview of the changes:
* Build bison and flex code as source files for the mod_auth_gssapi library instead of its own object. This removes some portability warnings about linking to mod_auth_gssapi. Use a separate header.
* Change the binary value specifier to ':=' (as in name:=value).
* Add the base64 value case to the parser where the incoming name attribute values are converted to base64 before comparing to the expected value.
* Store the environment name attributes and the to-be-verified name attributes separately in mag_conn, and prefer the display value.
* For testing generate x509 certificates and enable PKINIT with multiple PKINIT indicators, and add required attribute test locations. In order to avoid PKINIT messing with the other tests, add a new user (maguser3) and perform PKINIT only before running the required attribute tests.

One issue is that autotools seems to have some problem during libtool linking the standalone parser test program tests/natest.c, where it cannot resolve the apr_base64* functions used by the parser. I've tried quite a few different Makefile.am configurations for it with no luck, so some suggestions would be helpful. For the moment I've left those build statements commented out.

simo5 commented on README:
```diff
@@ -306,6 +307,33 @@ underscores for environment variable names.
+This option allows specifying one or more Name Attributes that the client must
+possess in order to be authorized to access the location. The required Name
+Attributes are specified by name=value pairs (name being the ATTRIBUTE_NAME as
+mentioned above, and value being an ASCII string. Alternately, if a Name
```
ASCII -> Can we say UTF8? Or do we have no guarntees in which case we should just use a "Null terminated string" ?

simo5 replied:
Now that I think of it, are spaces allowed in bare strings ?

mrogers950 replied:
I don't think there is a guarantee for UTF-8 here as I specified just character values in the lex rules (but it might be possible to specify UTF-8 code points instead). And spaces are not allowed in bare strings as that would be considered a token separation, so if there is a space expected in the value then the base64 option should be used. I can update the README with this clarification.

simo5 replied:
Yes, please, it will help nuderstanding when you must use bse64

simo5 commented on src/environ.c:
```diff
@@ -43,6 +43,64 @@ static apr_status_t mag_mc_name_attrs_cleanup(void *data)
+    for (count = 0; mc->required_name_attrs != NULL &&
+                    mc->required_name_attrs[count] != NULL &&
+                    mc->required_name_vals != NULL &&
+                    mc->required_name_vals[count] != NULL; count++);
```
Is it legal to have names w/o corresponding value ?

simo5 commented on src/mod_auth_gssapi.h:
```diff
@@ -128,6 +129,8 @@ struct mag_conn {
     bool is_preserved;
     int na_count;
+    const char **required_name_attrs;
+    const char **required_name_vals;
```
Looking at the code it seem like these two arrays are always allocated and handled in pairs.
Maybe we should have instead a:
struct mag_required_name_attr {
    const char *name;
    const char *value;
};

And just allocate an array of these ?
Given we often count them, we could also add a size_t req_name_attr_count, and keep their number around.

mrogers950 replied:
Right, they are a pair. Turning it into a struct with a counter for clarity is OK with me.

mrogers950 replied:
Actually, I opted to store them in this way to make it simple to pass to the parser. Storing them in a mag_attr struct would need a conversion later on (or a change in the parser). So I'd prefer to keep them this way and add in a comment for clarification.

frozencemetery commented:
I have no non-nit objections

frenche commented:
Great stuff, I plan to review some more over the weekend, (no need to hold the merge however).

frenche commented on src/environ.c:
```diff
@@ -43,6 +43,66 @@ static apr_status_t mag_mc_name_attrs_cleanup(void *data)
+        if (!mc->required_name_attrs || !mc->required_name_vals) {
+            apr_pool_abort_get(mc->pool)(ENOMEM);
+        }
+        apr_pool_userdata_setn(mc, GSS_NAME_ATTR_USERDATA,
```
Maybe better to be called once at mc constructor.

frenche replied:
Please ignore this comment, I now see it was modeled after the other directive, so let's leave it for now.

frenche commented on src/environ.c:
```diff
@@ -43,6 +43,66 @@ static apr_status_t mag_mc_name_attrs_cleanup(void *data)
+        val = apr_pstrndup(mc->pool, attr->display_value.value, len);
+    } else if (attr->value.length != 0) {
+        len = attr->value.length;
+        val = apr_pcalloc(mc->pool, len + 1);
```
If raw value may embed null, then the length must be kept along or it should be encoded here, otherwise it will get cut at the first null occurrence when accessed.

frenche commented on src/environ.c:
```diff
@@ -43,6 +43,66 @@ static apr_status_t mag_mc_name_attrs_cleanup(void *data)
+        len = attr->value.length;
+        val = apr_pcalloc(mc->pool, len + 1);
+        memcpy(val, attr->value.value, len);
+    }
```
Maybe else {error_out}, and perhaps move this block upper.

mrogers950 commented:
Updated, now we prepend the value length to the saved value. In the parser we read in the length prefix and use it for the value comparison. I moved the value checks in mag_set_required_name_attr() with an else return as suggested, and also stuck in an ap_log_rdata() call to print out the raw value data.

simo5 commented:
@mrogers950 github claims there are merge conflicts, can you check ?

mrogers950 commented:
Rebased to latest master.

simo5 commented:
@frenche can you confirm your remarks have been addressed ?

mrogers950 commented:
I added in the flex options to silence the remaining warnings.

simo5 commented:
make test fails for me:
{{{
cd . && ./tests/magtests.py --path /home/simo/devel/git/mod_auth_gssapi/testsdir --so-dir=/home/simo/devel/git/mod_auth_gssapi/src/.libs
/home/simo/devel/git/mod_auth_gssapi/testsdir/key.pem
SPNEGO: SUCCESS
SPNEGO Proxy Auth: SUCCESS
SPNEGO No Auth: SUCCESS
SPNEGO Rewrite: SUCCESS
SPNEGO Negotiate Once: SUCCESS
HOSTNAME ACCEPTOR: SUCCESS
BAD ACCEPTOR: SUCCESS
Traceback (most recent call last):
  File "./tests/magtests.py", line 722, in <module>
    testenv = kinit_certuser(testdir, testenv)
  File "./tests/magtests.py", line 477, in kinit_certuser
    raise ValueError('kinit failed')
ValueError: kinit failed
Makefile:821: recipe for target 'check' failed
make: *** [check] Error 1
}}}

Do we require some specific version of some dependency ?
If so can you add a check in configure ?

mrogers950 commented:
That happens when krb5-pkinit isn't installed, so I added a dirty check for the pkinit module and skip the name attribute tests with a warning if it's not found, and README.md mentions the krb5-pkinit dependency for the tests.
